src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    // PROTOTYPE(source-generators):
+    // the builder seems useful as a type, but the actual table is just a wrapper around a dict
+    // do we actually need the type itself, or should we just store the dict directly as
+    // part of the driver state?
+    internal sealed class DriverStateTable
+    {
+        // PROTOTYPE(source-generators): should we make a non generic node interface that we can use as the key
+        //                               instead of just object?
+        private readonly ImmutableDictionary<object, IStateTable> _tables;
+
+        internal static DriverStateTable Empty { get; } = new DriverStateTable(ImmutableDictionary<object, IStateTable>.Empty);
+
+        private DriverStateTable(ImmutableDictionary<object, IStateTable> tables)
+        {
+            _tables = tables;
+        }
+
+        public class Builder
+        {
+            private readonly ImmutableDictionary<object, IStateTable>.Builder _tableBuilder = ImmutableDictionary<object, IStateTable>.Empty.ToBuilder();
+
+            private readonly DriverStateTable _previousTable;
+
+            public Builder(DriverStateTable previousTable)
+            {
+                _previousTable = previousTable;
+            }
+
+            public NodeStateTable<T> GetLatestStateTableForNode<T>(IIncrementalGeneratorNode<T> source)
```

Comment:
What is the relationship between `IIncrementalGeneratorNode<T>` and an `IINcrementalGenerator<T>`?

Reply:
`IIncrementalGenerator` is the public API the user implements. Internally we create a graph of `IIncrementalGeneratorNode<T>`s for each call to ValueSourceExtensions.* in the pipeline callback